src/Oscoin/Storage/Block/Memory.hs

```diff
@@ -0,0 +1,58 @@
+module Oscoin.Storage.Block.Memory
+    ( newBlockStoreIO
+    ) where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Blockchain hiding (lookupTx)
+import           Oscoin.Crypto.Blockchain.Block (Block)
+import qualified Oscoin.Crypto.Hash as Crypto
+import qualified Oscoin.Storage.Block.Abstract as Abstract
+import qualified Oscoin.Time.Chrono as Chrono
+
+import qualified Data.List as List
+
+newBlockStoreIO
+    :: (MonadIO m, Crypto.Hashable c tx)
+    => Chrono.NewestFirst [] (Block c tx (Sealed c s))
+    -> IO (Abstract.BlockStore c tx s m)
+newBlockStoreIO blks = do
+    blocksVar <- newMVar $ Chrono.toNewestFirst blks
+    let getBlocks = liftIO $ readMVar blocksVar
+    let modifyBlocks f = liftIO $ modifyMVar_ blocksVar $ pure . f
+    pure $ mkBlockStore getBlocks modifyBlocks
+
+mkBlockStore
+    :: (Monad m, Crypto.Hashable c tx)
+    => m [Block c tx (Sealed c s)]
+    -> (([Block c tx (Sealed c s)] -> [Block c tx (Sealed c s)]) -> m ())
+    -> Abstract.BlockStore c tx s m
+mkBlockStore getBlocks modifyBlocks = (blockStoreReader, blockStoreWriter)
+  where
+    blockStoreReader = Abstract.BlockStoreReader
+        { getGenesisBlock =
+            List.last <$> getBlocks
+        , lookupBlock = \h ->
+            find (\b -> blockHash b == h) <$> getBlocks
+        , lookupTx = \txHash -> lookupTx txHash <$> getBlocks
+        , getBlocksByDepth = \d ->
+              Chrono.NewestFirst . take (fromIntegral d) <$> getBlocks
+        , getBlocksByParentHash = \h ->
+              Chrono.NewestFirst . takeWhile (\b -> blockHash b /= h) <$> getBlocks
+        , getTip = List.head <$> getBlocks
```

Comment:
Do you think we should use a `NonEmpty` list here, to avoid the use of unsafe functions?

Reply:
I replaced `[]` with `Seq` instead of using `NonEmpty`. But I’m requiring a `NonEmpty` blockchain when initializing the blockstore. This should enforce that we never have an empty blockstore.

Reply:
Right, but a `Seq` can itself be empty, no? Here I am arguing whether or not this is the right contract for this abstraction -- the blockchain must be non-empty by definition, so why not leverage that to get rid of annoying edge cases?

Any problems using `NonEmpty` here? Anything particular that you don't like?

Reply:
I just want to use `Seq` to get better performance. If you think that is not a concern I’m happy to switch to `NonEmpty`.

Reply:
On top of my head I don't think this is going to be a concern, and we can always switch to `Seq` later once we do a performance testing round. Better to start off with sound interfaces and limit total functions, that at least will be my preference 😉

Reply:
I tried to implement it with `NonEmpty`. Turns out that implementing `switchToFork` becomes a bit more involved. I’ll keep it as is (using `Seq`).